Added Jupyter Chapter 1.1 (Schnorr Introduction).

jachiang opened the pull request:
This is the first workshop chapter after an initial setup chapter (chaper 1.0).

An overview of the planned workshop chapters can be found [here](https://docs.google.com/spreadsheets/d/1c0aET8DugXarqSci72eufnjJFynd6xvKyVOkEzrhAhM/edit#gid=1256797271).

jachiang commented:
Ping @elichai for chapter review.

jachiang commented:
Update: Imports util.py

elichai commented on Solutions/1.1-Introduction-to-Schnorr-Solutions.ipynb:
```diff
@@ -0,0 +1,423 @@
+    "    \n",
+    "Whether a scalar is a quadratic residue of the secp256k1 field size is determined with the jacobi symbol:\n",
+    "\n",
+    "* Bitcoin Core method: `jacobi_symbol(k, SECP256K1_FIELD_SIZE) == 1`\n",
```
the Jacobi symbol isn't tested on `k`. it's tested on the Y of kG.
we want to know that kG produces a quad residue Y.

jnewbery replied:
fixed

elichai commented on Solutions/1.1-Introduction-to-Schnorr-Solutions.ipynb:
```diff
@@ -0,0 +1,423 @@
+    "\n",
+    "# Generate s = k + H(R_x|P|msg) * x\n",
+    "H_b = hashlib.sha256(R.get_bytes()[1:33] + P.get_bytes() + msg).digest()\n",
+    "H_int = int.from_bytes(H_b, 'big')\n",
```
mod group order?

jachiang replied:
Yes, that is better!

jnewbery replied:
fixed

elichai commented on Solutions/1.1-Introduction-to-Schnorr-Solutions.ipynb:
```diff
@@ -0,0 +1,423 @@
+    "\n",
+    "# Generate s = k + H(R_x|P|msg) * x\n",
+    "H_b = hashlib.sha256(R.get_bytes()[1:33] + P.get_bytes() + msg).digest()\n",
+    "H_int = int.from_bytes(H_b, 'big')\n",
```
mod group order?

jachiang replied:
Yes, that is better!

jnewbery replied:
fixed

elichai commented:
First of all wow. you did a lot of work :+1: 

Second, I hope we can try to simplify some of this. It's going to be pretty hard to get into the details you're getting even into more details and without confusing, the difference between a group order and field order for example is pretty confusing without trying to explain more theoretical parts.

Another thing is quadratic residue.
We are testing a quad residue over `y^2=x^3+7` so the quad residue is on `y` over the field order `p`.
so either quadratic residue in the curve equation. or just `quadratic residue mod p` or even just `quadratic residue`.

jachiang commented on 1.1-Introduction-to-Schnorr.ipynb:
```diff
@@ -0,0 +1,388 @@
+    "    * For x, both P(x,y) and P(x,-y) are valid curve points.\n",
+    "    * One of the y-coordinates is even, and the other one is odd.\n",
+    "\n",
+    "The nonce k is therefore constrained by requiring it to be a quadratic residue of p (SECP256k1 curve field size). \n",
```
Right, but doesn't this infer a constraint on k? As we negate(k) if the constraint is not met..

elichai replied:
yes it does, but you cannot check this constraint using `k` only.

jnewbery replied:
fixed

jachiang commented on 1.1-Introduction-to-Schnorr.ipynb:
```diff
@@ -0,0 +1,388 @@
+    "The nonce k is therefore constrained by requiring it to be a quadratic residue of p (SECP256k1 curve field size). \n",
+    "\n",
+    "* `k` and its complement `SECP26k1_order - k` will have nonce points `R(x, y)` and `R(x, -y)`.\n",
+    "    * Only one will be a quadratic residue of the curve order.\n",
```
Ah, yes you are right, modulo the field order, thanks!

jnewbery replied:
fixed

jachiang commented:
I have now updated this PR and split out the EC-math section to chapter 0.2 (EC math review), as discussed in our call previously.

@elichai Many thanks for your review. Your corrections have been incorporated, with one [exception](https://github.com/bitcoinops/taproot-workshop/pull/13#issue-313677062).

@bitschmidty Appreciate your critical review of this chapter and [0.2 ec math review](https://github.com/bitcoinops/taproot-workshop/pull/13#issue-313677062).

jnewbery commented:
rebased on master

jnewbery commented:
@elichai - thanks for reviewing! I like your suggestions about making ECKey and ECPubKey easier to deal with. Are you able to make a PR to do that? We can add new methods to ECKey and ECPubKey in util.py and then import from there.

jachiang commented:
Updated with input from feedback. Incorporates __add__/__mul__ key class methods in code solutions.

jnewbery commented on 1.1-Introduction-to-Schnorr.ipynb:
```diff
@@ -0,0 +1,388 @@
+    "### Classes / Methods for Elliptic Curve Math\n",
+    "\n",
+    "`int`: All Scalar operations over SECP251k1 can be performed with python integers and the modulo `%` operator:\n",
+    "* Addition: `a + b % SECP256K1`\n",
```
resolved in the latest 0.2 notebook

jnewbery commented on 1.1-Introduction-to-Schnorr.ipynb:
```diff
@@ -0,0 +1,388 @@
+    "\n",
+    "# ECkey(operation)\n",
+    "a_key = ECKey()\n",
+    "a_key.set(a.to_bytes(32,'big'), True) \n",
```
Tracking here: https://github.com/bitcoinops/taproot-workshop/issues/28

jnewbery commented on 1.1-Introduction-to-Schnorr.ipynb:
```diff
@@ -0,0 +1,388 @@
+   "source": [
+    "# int() operations\n",
+    "a = random.randrange(1, SECP256K1_ORDER)\n",
+    "b = random.randrange(1, SECP256K1_ORDER)\n",
```
I think what we have now is ok, but feel free to open a PR if you think it can be improved.

jnewbery commented on 1.1-Introduction-to-Schnorr.ipynb:
```diff
@@ -0,0 +1,388 @@
+    "# We have to set \"compressed\" to false in ECKey.set(data, False) to get the y-coordinate.\n",
+    "\n",
+    "\n",
+    "# Check that nonce is quadratic residue of field size.\n",
```
fixed

jnewbery commented on 1.1-Introduction-to-Schnorr.ipynb:
```diff
@@ -0,0 +1,388 @@
+    "\n",
+    "\n",
+    "# Check that nonce is quadratic residue of field size.\n",
+    "# Method: jacobi_symbol(nonce_int, SECP256K1_FIELD_SIZE)\n",
```
I think it's fine to keep this explicitly as `SECP256K1_FIELD_SIZE`

jnewbery commented:
I think all feedback in this PR is addressed in #26
